File: android/app/src/main/java/uk/co/amlcurran/social/JodaCalculator.java
package uk.co.amlcurran.social;

import org.jetbrains.annotations.NotNull;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.Days;

public class JodaCalculator implements TimeCalculator {
    private static final DateTime EPOCH = new DateTime(0, DateTimeZone.getDefault());

    @NotNull
    @Override
    public Timestamp plusHours(Timestamp time, int hours) {
        return new Timestamp(getDateTime(time).plusHours(hours).getMillis(), this);
    }

    @Override
    public Timestamp startOfToday() {
        return new Timestamp(DateTime.now().withTimeAtStartOfDay().getMillis(), this);
    }

    @Override
    public int getDays(Timestamp time) {
        return Days.daysBetween(EPOCH, getDateTime(time)).getDays();
    }

    @NotNull
    @Override
    public Timestamp plusDays(int days, Timestamp time) {
        return new Timestamp(getDateTime(time).plusDays(days).getMillis(), this);
    }

    public DateTime getDateTime(Timestamp time) {
        return new DateTime(time.getMillis(), DateTimeZone.getDefault());
    }
}

File: android/app/src/main/java/uk/co/amlcurran/social/JodaCalculator.kt
package uk.co.amlcurran.social

import kotlinx.datetime.Instant
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import org.joda.time.Days

class JodaCalculator : TimeCalculator {

    override fun plusHours(time: Timestamp, hours: Int): Timestamp {
        return Instant.fromEpochMilliseconds(getDateTime(time).plusHours(hours).millis)
    }

    override fun startOfToday(): Timestamp {
        return Instant.fromEpochMilliseconds(DateTime.now().withTimeAtStartOfDay().millis)
    }

    override fun getDays(time: Timestamp): Int {
        return Days.daysBetween(EPOCH, getDateTime(time)).days
    }

    override fun plusDays(days: Int, time: Timestamp): Timestamp {
        return Instant.fromEpochMilliseconds(getDateTime(time).plusDays(days).millis)
    }

    fun getDateTime(time: Timestamp): DateTime {
        return DateTime(time.toEpochMilliseconds(), DateTimeZone.getDefault())
    }

    companion object {
        private val EPOCH = DateTime(0, DateTimeZone.getDefault())
    }
}
